### PR TITLE
Add system data streams to feature state snapshots

### LINKED ISSUE #75860: Include system data streams when snapshotting feature states
In the course of work on #75819, we discovered that we have not connected system data streams and snapshots. If a user creates a snapshot for a specific set of indices and the Fleet feature state, the fleet data stream and backing indices will not be included. This is a bug, though at the moment it only affects Fleet. (A general snapshot and restore of the cluster will include all feature states.)

We need to add system data stream handling to the snapshot and restore code for feature states.

### PR DESCRIPTION
**Pull request by williamrandolph**
This PR adds system data streams to the "snapshot feature state" code block, so that if we're snapshotting a feature by name we grab that feature's system data streams too.

Fixes #75860

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (gwbrown) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:349]: Nitpick: I think we should either clarify this to something like `Include both requested data streams as well as system data streams included via feature states`, or just remove this comment. As is, I don't think this adds much and risks being mistaken for a `TODO`.

File: x-pack/plugin/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/SystemDataStreamSnapshotIT.java
Comment (gwbrown) [inline x-pack/plugin/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/SystemDataStreamSnapshotIT.java:176]: Added a comment to the linked issue so we remember to clean up the places we have to work around this in tests. No action required here, just noting.

**Review by AthenaEryma — APPROVED** [review]
LGTM apart from a couple nitpicks, but I think it'd be a good idea to also get a +1 from the Distributed team - I'll add someone.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:340]: Could we have duplicate index names/data-streams in these lists (like `featureSystemIndices`)? I'm mainly asking because on the snapshot side we're handling everything as `List` while when restoring we do create sets which only really add anything if there's duplicates so either we can go with just lists on the restore end or we should use sets here?

**Review by original-brownbear — COMMENTED** [review]
Just a smaller question and test cleanup suggestions, looking good otherwise.

File: server/src/main/java/org/elasticsearch/indices/SystemDataStreamDescriptor.java
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/indices/SystemDataStreamDescriptor.java]: I think it was that I didn't want to use the non-static method `getBackingIndexPattern()` in the constructor, but I've refactored a little bit to get around that.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:342]: I will take a look at turning a few of these Lists into Sets:
- [ ] `featureStates`, `systemDataStreamNames`
- [ ] Other collections used only within the for-loop

Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: @original-brownbear I think I've addressed all the issues you raised. Do you think we're good to merge this one?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:342]: There's a bit of a bug here, that's not new but more explicit now in that we create `SnapshotFeatureInfo` (something living in the CS) with a mutable list field. I'd just use `List.copyOf` here instead of `new ArrayList` for now to fix this spot.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:399]: Let's use List.copyOf() here as well, no need to mutable lists as ctor arguments here.

File: x-pack/plugin/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/SystemDataStreamSnapshotIT.java
Comment (original-brownbear) [reply x-pack/plugin/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/SystemDataStreamSnapshotIT.java:209]: This method seems unused now?

**Review by original-brownbear — COMMENTED** [review]
Thanks @williamrandolph looks good for the most part :) Just a few small points that I found on re-reading this and one new question.

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:336]: That's a great catch. Thank you!

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:340]: Use normal style logging with placeholders instead of concatenation?

**Review by original-brownbear — APPROVED** [review]
LGTM, if CI is happy I'm happy, though maybe fix the logging :)

**Comment by williamrandolph** [comment]
Backport PR: #76568

**Pull request merged**
